ProcessList.c

```diff
@@ -112,16 +115,73 @@ int ProcessList_size(ProcessList* this) {
    return (Vector_size(this->processes));
 }
 
-static void ProcessList_buildTree(ProcessList* this, pid_t pid, int level, int indent, int direction, bool show) {
+static int ProcessList_updateTreeSetLayer(ProcessList* this, unsigned int deep, unsigned int lft, unsigned int rht, int* index, int* tree_index, int indent) {
+   Vector* layer = Vector_new(this->processes->type, false, DEFAULT_SIZE);
+   for (int i = 0; i < Vector_size(this->processes); i++) {
+      Process* proc = (Process*)Vector_get(this->processes, i);
+      if (proc->tree_deep == deep && proc->tree_lft > lft && proc->tree_rht < rht) {
+         Vector_add(layer, proc);
+      }
+   }
+
+   Vector_quickSort(layer);
+
+   int size = Vector_size(layer);
+   for (int i = 0; i < Vector_size(layer); i++) {
+      int lft = (*tree_index)++;
+      int idx = (*index)++;
+
+      int level = deep-1;
+      int currentIndent = indent == -1 ? 0 : indent | (1 << level);
+      int nextIndent = indent == -1 ? 0 : (i < size - 1) ? currentIndent : indent;
+
+      Process* proc = (Process*) (Vector_get(layer, i));
+      size += ProcessList_updateTreeSetLayer(this, deep+1, proc->tree_lft, proc->tree_rht, index, tree_index, nextIndent);
+
+      int rht = (*tree_index)++;
+
+      Hashtable_put(this->treeSet, idx, proc);
+      proc->tree_lft = lft;
+      proc->tree_rht = rht;
+
+      if (indent == -1) {
+         proc->indent = 0;
+      } else if (i == size - 1)
+         proc->indent = -currentIndent;
+      else
+         proc->indent = currentIndent;
+   }
+
+   Vector_delete(layer);
+
+   return size;
+}
+
+static void ProcessList_updateTreeSet(ProcessList* this) {
```

Comment:
Some short documentation what the various new functions do and how they interact might be of help. E.g. what's the difference between `ProcessList_updateTreeSet` and `ProcessList_buildTree`. What are some pre-conditions these functions expect?

Reply:
A good question. Both of these build a treeSet.

But `ProcessList_buildTree` doesn't use a privious map so basically it does look up the children by pid. Where `ProcessList_updateTreeSet` rebuild map only on knowledge of map.

https://github.com/htop-dev/htop/blob/a0c24b809dd7970019148b12d34c998f62b33bdc/ProcessList.c#L176-L182

https://github.com/htop-dev/htop/blob/a0c24b809dd7970019148b12d34c998f62b33bdc/ProcessList.c#L120-L125

That's the whole difference I suppose